docs/design_2x.md

```diff
@@ -1,34 +1,130 @@
-synthetic application abstraction models in place of real application code.
-
-What about mininet? [mininet](http://mininet.org) is a network emulator that is
-designed to run real kernel, switch, and application code in real time. The real
-time requirement severely limits the number of processes that can be run in a
-mininet experiment: time distortion can occur if the processes exceed a
-computational threshold, which can result in undefined behavior and artifacts
-that lead to untrustworthy results.
-
-Shadow aims to fill the gap between these tools. Like mininet, Shadow directly
-executes real applications in order to faithfully reproduce application
-behavior. But like ns-3, Shadow runs a discrete-event simulation in order to
-scientifically control and deterministically replicate network experiments.
-Shadow uniquely targets experiments with large-scale distributed systems and
-thus our simulator design prioritizes high-performance computing.
+# Shadow Design Overview
+
+Shadow is a multi-threaded network experimentation tool that is designed as a
+hybrid between simulation and emulation architectures: it directly executes
+applications as Linux processes, but runs them in the context of a
+discrete-event network simulation.
+
+Shadow's version 2 design in summarized in the following sections. Please see
+the end of this document for references to published design articles with more
+details.
+
+## Executing Applications
+
+Shadow directly executes real, unmodified application binaries natively in Linux
+as standard OS processes (using `vfork()` and `execvpe()`): we call these
+processes executed by Shadow _managed processes_. When executing each managed
+ process, Shadow dynamically injects a shim library using preloading (via the
+`LD_PRELOAD` environment variable) and establishes an inter-process control
+channel using shared memory and semaphores. The control channel enables Shadow
+to exchange messages with the shim and to instruct the shim to perform actions
+in the managed process space.
+
+## Intercepting System Calls
+
+The shim co-opts each running managed process into the simulation environment by
+intercepting all system calls they make rather than allowing them to be handled
+by the Linux kernel. System call interception happens through two methods: first
+via preloading and second via a seccomp filter.
+
+- Preloading: Because the shim is preloaded, the shim will be the first library
+that is searched when attempting to dynamically resolve symbols. We use the shim
+to override functions in other shared libraries (e.g., system call wrapper
+functions from libc) by supplying identically named functions with alternative
+implementations inside the shim. Note that preloading works on dynamically
+linked calls to libc system call wrappers, but not on statically linked calls
+made from inside of libc or made using a `syscall` instruction.
+
+- seccomp: System calls that are not preloadable are intercepted using the
```

Comment:
Maybe "interceptible via preloading" instead of "preloadable". Or if we think it's really worth inventing a term, maybe define it in the previous paragraph. "e.g. We call system calls that can be intercepted this way preloadable"